File: App/pidrei-ng-app/src/shared/classes/thermostat.ts
import { ThermostatChange } from './thermostat-change';

export class Thermostat {
    constructor() {
        this.nextChange = new ThermostatChange();
    }
    public actualTemperature: number;
    public nominalTemperature: number;
    public savingTemperature: number;
    public comfortTemperature: number;
    public isUiLocked: Boolean;
    public isDeviceLocked: Boolean;
    public errorCode: number;
    public isBatteryLow: Boolean;
    public nextChange: ThermostatChange;

    public assign(other: Thermostat): Thermostat {
        this.actualTemperature = other.actualTemperature;
        this.nominalTemperature = other.nominalTemperature;
        this.savingTemperature = other.savingTemperature;
        this.comfortTemperature = other.comfortTemperature;
        this.isUiLocked = other.isUiLocked;
        this.isDeviceLocked = other.isDeviceLocked;
        this.errorCode = other.errorCode;
        this.isBatteryLow = other.isBatteryLow;
        this.nextChange.changeTime = other.nextChange.changeTime;
        this.nextChange.newTemperature = other.nextChange.newTemperature;

        return this;
    }

    public isActivated(): boolean {
        return this.nominalTemperature !== -273.15;
    }
}

File: FritzBox/Model/Thermostat.cs
using System;
using System.Xml.Serialization;
using Newtonsoft.Json;

namespace RaspPiTest.FritzBox.Model
{
    [JsonObject(MemberSerialization.OptIn)]
    public class Thermostat
    {
        [XmlElement("tist")]
        public byte ActualTemperatureInternal { get; set; }
        [XmlElement("tsoll")]
        public byte NominalTemperatureInternal { get; set; }
        [XmlElement("absenk")]
        public byte SavingTemperatureInternal { get; set; }
        [XmlElement("komfort")]
        public byte ComfortTemperatureInternal { get; set; }
        [XmlElement("lock")]
        public byte? IsUiLockedInternal { get; set; }
        [XmlElement("devicelock")]
        public byte? IsDeviceLockedInternal { get; set; }
        [XmlElement("errorcode")]
        public byte ErrorCodeInternal { get; set; }
        [XmlElement("batterylow")]
        public byte IsBatteryLowInternal { get; set; }
        [XmlElement("nextchange")]
        [JsonProperty("nextChange")]
        public ThermostatChange NextChange { get; set; }

        public static float GetTemperature(byte internalValue)
        {
            if (internalValue == 253) return -273.15f;
            if (internalValue == 254) return 0;
            return internalValue / 2f;
        }

        public static byte GetFritzboxTemperature(float celsiusTemperature)
        {
            if (celsiusTemperature < 0) return 253;
            if (Math.Abs(celsiusTemperature) < 0.1) return 254;
            return Convert.ToByte(2 * celsiusTemperature);


        }

        [JsonProperty("actualTemperature")]  public float ActualTemperature => GetTemperature(ActualTemperatureInternal);
        [JsonProperty("nominalTemperature")] public float NominalTemperature => GetTemperature(NominalTemperatureInternal);
        [JsonProperty("savingTemperature")]  public float SavingTemperature => GetTemperature(SavingTemperatureInternal);
        [JsonProperty("comfortTemperature")] public float ComfortTemperature => GetTemperature(ComfortTemperatureInternal);
        [JsonProperty("isUiLocked")]         public bool IsUiLocked => IsUiLockedInternal == 1;
        [JsonProperty("isDeviceLocked")]     public bool IsDeviceLocked => IsDeviceLockedInternal == 1;
        [JsonProperty("errorCode")]          public ThermostatErrorCode ErrorCode => (ThermostatErrorCode) ErrorCodeInternal;
        [JsonProperty("isBatteryLow")]       public bool IsBatteryLow => IsBatteryLowInternal == 1;
    }
}